### PR TITLE
LLM Chat Interface (& RAG) Integration into LW

### PR DESCRIPTION
**Pull request by darkruby501**
(before I get to the main PR, want to flag for EA Forum folks @jpaddison3:)

We added typing to the local storage handling and then had to make fixes elsewhere for the type errors. We believe this should be fine but flagging fyi.

-----------
This PR sets up LessWrong to have an LLM chat integration with Claude (3.5 Sonnet, though that's easily changed).

LessWrong 2024-08-26 14-32-03

LessWrong 2024-08-26 14-37-13

LessWrong 2024-08-26 14-40-43

Various facts:
- The Claude API is stateless, for every message submitted by the user to us, we submit the entire conversation history to Claude API
- There are collections for `LLMConversations` and `LlmMessages`
- Currently only admins and manually added users (there's a DB setting to grant access to non-admin users).
- In addition to the main calls to Claude Sonnet 3.5 for conversation responses, we make small quick calls to faster models for things like automatically generating titles and helping select which additional context to load.

### Retrieval Augmented Generation (adding LW context to the chat context)
The major motivations for building this were (1) we'll possibly be able to prototype many different ideas on top of this, and (2) this allows for easily adding LW posts as context.

This first version uses a primitive embedding search based on the user's opening query and/or post currently being viewed to select posts to feed to the LLM with appropriate prompting. Comments are included for the post currently being viewed.

The context-selection and loading is currently so-so and we'll iterate on it after getting the basic thing shipped.

TO-DO:
It's possible to rack up non-trivial expenses using the Claude API with large amounts of context. Before rolling out generally, some kind of payment system or users submitting their own API keys is necessary.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1207985488494994) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/components/languageModels/LanguageModelChat.tsx [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+  userMessage: {
+    backgroundColor: theme.palette.grey[300],
+  },
+  messages: {
+    overflowY: "scroll",
+    flexGrow: 1,
+    //the last child has margin bottom 0 (doesn't work)
+    "& > *:last-child": {
+      marginBottom: 0
+    }
```
Comment (b0b3rt) [inline packages/lesswrong/components/languageModels/LanguageModelChat.tsx:90]: This doesn't work?

File: packages/lesswrong/components/languageModels/LanguageModelLauncherButton.tsx [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    display: "flex",
+    justifyContent: "center",
+    alignItems: "center",
+    borderRadius: "50%",
+    width: 48,
+    height: 48,
+    padding: 10,
+    cursor: "pointer",
+    // TODO: Figure out for dark mode
+    // backgroundColor: theme.palette.grey[100],
```
Comment (b0b3rt) [inline packages/lesswrong/components/languageModels/LanguageModelLauncherButton.tsx:23]: ^

File: packages/lesswrong/components/languageModels/LanguageModelLauncherButton.tsx [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    boxShadow: "0 1px 6px 0 rgba(0, 0, 0, 0.06), 0 2px 32px 0 rgba(0, 0, 0, 0.16)",
+    // add hover styling
+    "&:hover": {
+      boxShadow: "0 6px 8px rgba(0, 0, 0, 0.2)"
```
Comment (b0b3rt) [inline packages/lesswrong/components/languageModels/LanguageModelLauncherButton.tsx:27]: Why the comment?

File: packages/lesswrong/components/languageModels/PopupLanguageModelChat.tsx [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  //   height: 20,
+  //   position: "absolute",
+  //   right: 32,
+  //   top: 10,
+  //   cursor: "pointer",
+  //   color: theme.palette.grey[400],
+  //   '&:hover': {
+  //     color: theme.palette.grey[600],
+  //   }
+  // },
```
Comment (b0b3rt) [inline packages/lesswrong/components/languageModels/PopupLanguageModelChat.tsx:52]: cleanup?

File: packages/lesswrong/lib/betas.ts [hunk 5]
```diff
@@ -74,6 +74,12 @@ export const useRecombeeFrontpage = (currentUser: UsersCurrent|DbUser|null) => {
   return isLW && (isAdmin(currentUser) || manualOptIn) && recombeeEnabledSetting.get()
 }
 
+export const userHasLlmChat = (currentUser: UsersCurrent | DbUser): boolean => {
+  const userIdsWithAccess = userIdsWithAccessToLlmChat.get();
+
+  return isLW && (isAdmin(currentUser) || userIdsWithAccess.includes(currentUser?._id));
```
Comment (b0b3rt) [inline packages/lesswrong/lib/betas.ts:80]: No point in doing a null-safety check if `currentUser` can't be null or undefined.  (But you probably should let `currentUser` be null in the function signature; that's what we do elsewhere and it's less annoying to use `userHasLlmChat` if you don't need to check whether `currentUser` exists first.)

File: packages/lesswrong/components/languageModels/LanguageModelChat.tsx (hunk 1 above)
Comment (darkruby501) [reply packages/lesswrong/components/languageModels/LanguageModelChat.tsx:89]: Empirically not

File: packages/lesswrong/components/languageModels/LanguageModelLauncherButton.tsx (hunk 2 above)
Comment (darkruby501) [reply packages/lesswrong/components/languageModels/LanguageModelLauncherButton.tsx:23]: I've edited the comment. The remaining stuff that doesn't work on darkmode is the shadow and hover effect

File: packages/lesswrong/components/languageModels/LanguageModelLauncherButton.tsx (hunk 3 above)
Comment (darkruby501) [reply packages/lesswrong/components/languageModels/LanguageModelLauncherButton.tsx:27]: I don't know. Removed

**Review by b0b3rt — APPROVED** [review]

**Pull request merged**